### PR TITLE
Fixing bug in decoder eval for optional fields

### PR DESCRIPTION
**Pull request by rpmcginty**
Issue: https://github.com/lidatong/dataclasses-json/issues/353

### Problem

There is inconsistent behavior in how a custom decoder is used depending on whether the field of the custom decoder has `Optional` type. 

For optional fields which can be *None*, you are supposed to use the `Optional[...]` type. However, when doing so, it seems that custom decoders are **always** evaluated.    

It should be noted that fields where the default value is *None* but no `Optional[...]` typing is present, a custom decoder will only be used if a non-None value is provided.   

### Solution

Modified decoder function s.t. when field value is None, both cases where field type is not optional (existing logic) as well as where field type is optional are handled.

### CODE REVIEW

**Comment by rpmcginty** [comment]
@lidatong is there any concern with this approach?

**Review by lidatong — APPROVED** [review]

**Review by lidatong — APPROVED** [review]

**Comment by rpmcginty** [comment]
@lidatong is there a timeline for merging into the package?

**Comment by TristanSpeakEasy** [comment]
@lidatong we have run into this issue as well, can we get this merged and a new release created?

**Comment by simplesagar** [comment]
@lidatong Ran into this issue as well. Would love to know if its getting merged soon :)

**Comment by patricksnape** [comment]
Whilst we are piling on - I would also love to see this merged, this causes very surprising results when a field is constructed incorrectly by passing None where None is not valid (you essentially just get a warning and then are provided with a broken class back.

**Comment by TristanSpeakEasy** [comment]
Unfortunately it seems that this project is likely abandoned, we have been maintaining a separate fork for our own use-case but would love to see this repo get some love @lidatong or a new maintainer take up the reigns (which might need to be via a fork of this if the original author doesn't respond)

**Comment by lidatong** [comment]
> Unfortunately it seems that this project is likely abandoned, we have been maintaining a separate fork for our own use-case but would love to see this repo get some love @lidatong or a new maintainer take up the reigns (which might need to be via a fork of this if the original author doesn't respond)

hi, i'm really sorry for the long delay. it's been a combination of work stress and burnout, but that is no excuse for my procrastination. i do really appreciate PRs from the community and will try to review and merge the high-demand ones this weekend! i'm also open to granting write permissions to repeat contributors, to make this project more sustainable

**Comment by rpmcginty** [comment]
Is there an option for the latter?

**Comment by rpmcginty** [comment]
@lidatong Let us know if granting wire permissions are an option. It would be good to get a community to help you out with making updates to this really useful package

**Comment by lidatong** [comment]
> @lidatong Let us know if granting wire permissions are an option. It would be good to get a community to help you out with making updates to this really useful package

sorry, i just fixed CI issues. definitely can grant write, though publishing is a final step i'll aim to not bottleneck -- it would be good to for to review all changes anyways before signing anything

**Review by lidatong — APPROVED** [review]

File: tests/test_metadata.py [hunk 1]
```diff
@@ -39,3 +51,12 @@ def test_mm_field_takes_precedence_over_types(self):
...
         assert obj.string_date == 'yesterday'
+
+    def test_optional_field_only_decoded_when_present(self):
+        obj = opt_string_date_schema.load({})
+        assert isinstance(obj, OptionalStringDate)
+        assert obj.string_date == None
+        
+        another_obj = opt_string_date_schema.load({'string_date': 'today'})
+        assert isinstance(another_obj, OptionalStringDate)
+        assert another_obj.string_date == 'today'
```
Comment (matt035343) [inline tests/test_metadata.py:62]:
```suggestion
        assert another_obj.string_date == 'today'

```

File: dataclasses_json/core.py [hunk 2]
```diff
@@ -158,17 +158,18 @@ def _decode_dataclass(cls, kvs, infer_missing):
...
-            if infer_missing:
-                warnings.warn(
-                    f"Missing {warning} and was defaulted to None by "
-                    f"infer_missing=True. "
-                    f"Set infer_missing=False (the default) to prevent this "
-                    f"behavior.", RuntimeWarning)
-            else:
-                warnings.warn(f"`NoneType` object {warning}.", RuntimeWarning)
+        if field_value is None:
+            if not _is_optional(field_type):
```
Comment (matt035343) [inline dataclasses_json/core.py:162]: Any particular reason to split the if-statement and introduce further nesting?

**Comment by matt035343** [comment]
Also, please sync your branch to latest master commit. Then we can proceed with a merge

File: dataclasses_json/core.py (hunk 2 above)
Comment (rpmcginty) [reply dataclasses_json/core.py:162]: The line at 172 (old code) `init_kwargs[field.name] = field_value` is only hit if the value is None *and* `field_type` is not optional. If anything, it should update `init_kwargs` when `field_type` is optional.

**Comment by rpmcginty** [comment]
@matt035343 I have rebased and fixed formatting errors

File: dataclasses_json/core.py (hunk 2 above)
Comment (matt035343) [reply dataclasses_json/core.py:171]: Apologies, I missed the part that `init_kwargs[field.name] = field_value` was not indented as well.

**Comment by george-zubrienko** [comment]
> Unfortunately it seems that this project is likely abandoned, we have been maintaining a separate fork for our own use-case but would love to see this repo get some love @lidatong or a new maintainer take up the reigns (which might need to be via a fork of this if the original author doesn't respond) 

No stress, there are extra hands on deck now :) please let me or @s-vitaliy or @matt035343 know if there are other PRs you want reviewed asap

**Pull request merged**